src/codegen/compile.ml

```diff
@@ -1602,12 +1608,30 @@ module Tagged = struct
     assert (!Flags.gc_strategy = Flags.Incremental);
     1l
 
-  (* Note: Post allocation barrier must be applied after initialization *)
+  (* Note: post-allocation barrier must be applied after initialization *)
   let alloc env size tag =
+    assert (size > 1l);
     let name = Printf.sprintf "alloc_size<%d>_tag<%d>" (Int32.to_int size) (Int32.to_int (int_of_tag tag)) in
+    let overflow_mask n =
+      let n = Int32.to_int n in
+      Int32.(logand 0xFFFFl (shift_left minus_one (16 - Numerics.Nat16.(to_int (clz (of_int n)))))) in
+
     Func.share_code0 env name [I32Type] (fun env ->
-      let (set_object, get_object) = new_local env "new_object" in
-      Heap.alloc env size ^^
+      let set_object, get_object = new_local env "new_object" in
+      let size_in_bytes = Int32.(mul size Heap.word_size) in
+      (if !Flags.gc_strategy <> Flags.Incremental && size_in_bytes < 0x8000l then
+         GC.get_heap_pointer env ^^
+         compile_sub_const 1l ^^ (* skew *)
+         GC.get_heap_pointer env ^^
+         compile_add_const size_in_bytes ^^
+         GC.set_heap_pointer env ^^
+         GC.get_heap_pointer env ^^
+         compile_bitand_const (overflow_mask size_in_bytes) ^^
+         G.if0
+           G.nop (* no page crossing *)
+           (Heap.alloc env 0l ^^ G.i Drop) (* ensure that the page is allocated *)
+       else
+         Heap.alloc env size) ^^
```

Comment:
If I am not mistaken, this case may also happen if the page is not necessarily crossed, e.g. for the first few small allocations in the page, e.g. for an allocation series of 32 bytes, 64 bytes, 128 bytes? Then, the leading zeros of the bump pointer inside the page remains 0?

Reply:
This line is for sizes that are too big...
Assume HP is at the page boundary (`0xPPPP0000`), then an 8-byte alloc would give `0xPPPP0008` which gives a false positive for the current mask (entering the slow path), but would be on the fast path with an extended mask. I considered this optimisation, but never proved it correct. I'll put this on the potential refinements list, as I don't want to change this PR any more.

Reply:
Thanks for the explanation.

Reply:
Okay, let's assume we can extend the mask. Increment
- `0b0000000001111100` (=`0x4c` = 76), augmented mask
- `0b11111111111111000000` (= `0xffc0`)
now suppose the old `HP` is `0x3333fffc`, then the post-bump `HP` is
`0x33340048`. Masking gives `0x0040` saying that no crossing happened, which is clearly wrong.
We can only do the mask extension when the bit after the topmost `1` in the increment is clear!

Reply:
Yes, the bit mask cannot be simply extended.
I assume the version with skewed pointer is working because we use skew = -1. 
If we would encode skew = +1, then it would not work?

Reply:
The mask extension is only possible for `popcnt N == 1` and also works for skewed `HP`s!